front/src/controlled_input.tsx

```diff
@@ -0,0 +1,222 @@
+    const [included, setIncluded] = useState(props.initialIncluded);
+    const [excluded, setExcluded] = useState(props.initialExcluded);
+
+    useDeepEffect(() => {
+        setIncluded(props.initialIncluded);
+        setExcluded(props.initialExcluded);
+    }, [props.initialIncluded, props.initialExcluded]);
+
+    const transfer = (value: number) => () => {
+        const newIncluded = new Set(included);
+        const newExcluded = new Set(excluded);
+
+        if (newIncluded.delete(value)) {
+            newExcluded.add(value);
+        } else {
+            newExcluded.delete(value);
+            newIncluded.add(value);
+        }
+
+        setIncluded(newIncluded);
+        setExcluded(newExcluded);
+        props.onChange(Array.from(newIncluded));
+    };
+
+    const includeAll = () => {
+        const newIncluded = new Set([...included, ...excluded]);
+        setIncluded(newIncluded);
+        setExcluded(new Set());
+        props.onChange(Array.from(newIncluded));
+    };
+    const excludeAll = () => {
+        setIncluded(new Set());
+        setExcluded(new Set([...included, ...excluded]));
+        props.onChange([]);
+    };
+
+    const customList = (title: string, items: Set<any>, is_left: boolean) => {
+        const ordered_items = Array.from(items);
+        ordered_items.sort((a, b) => primitive_to_str(a).localeCompare(primitive_to_str(b)));
+        return (
+            <Card>
+                <CardHeader sx={{ px: 2, py: 1 }} title={title} />
+                <Divider />
+                <List dense component='div' role='list'>
+                    {ordered_items.map((value: any) => {
+                        const labelId = `transfer-list-item-${value}-label`;
+
+                        return (
+                            <ListItem key={value} role='listitem' button onClick={transfer(value)}>
+                                <ListItemIcon>
+                                    {is_left ? <ArrowCircleRightIcon /> : <ArrowCircleLeftIcon />}
+                                </ListItemIcon>
+                                <ListItemText id={labelId} primary={primitive_to_str(value)} />
+                            </ListItem>
+                        );
+                    })}
+                    <ListItem />
+                </List>
+            </Card>
+        );
+    };
+
+    return (
+        <Grid container spacing={2} justifyContent='center' alignItems='center'>
+            <Grid item>{customList('Included', included, true)}</Grid>
+            <Grid item>
+                <Grid container direction='column' alignItems='center'>
+                    <Button
+                        sx={{ my: 0.5 }}
+                        variant='outlined'
+                        size='small'
+                        onClick={excludeAll}
+                        disabled={included.size === 0}
+                        aria-label='exclude all'
+                    >
+                        ≫
```

Comment:
you can use material icons 😊

Reply:
I actually copied this form MUI's example https://mui.com/material-ui/react-transfer-list/#basic-transfer-list